packages/rum-core/src/domain/foregroundContexts.ts

```diff
@@ -0,0 +1,123 @@
+  stop: () => void
+}
+
+export interface FocusPeriod {
+  start: RelativeTime
+  end?: RelativeTime
+}
+
+let focusPeriods: FocusPeriod[] = []
+
+export function startForegroundContexts(configuration: Configuration) {
+  if (!configuration.isEnabled('track-foreground')) {
+    return {
+      getInForeground: () => undefined,
+      getInForegroundPeriods: () => undefined,
+      stop: noop,
+    }
+  }
+  if (document.hasFocus()) {
+    addNewFocusPeriod()
+  }
+
+  const { stop: stopFocusTracking } = trackFocus(addNewFocusPeriod)
+  const { stop: stopBlurTracking } = trackBlur(closeFocusPeriod)
+  return {
+    getInForeground,
+    getInForegroundPeriods,
+    stop: () => {
+      focusPeriods = []
+      stopFocusTracking()
+      stopBlurTracking()
+    },
+  }
+}
+
+function addNewFocusPeriod() {
+  const now = relativeNow()
+  if (focusPeriods.length > MAX_NUMBER_OF_FOCUSED_TIME) {
+    return
+  }
+  const lastIndex = focusPeriods.length - 1
+  if (lastIndex >= 0 && focusPeriods[lastIndex].end == null) {
+    // already here
+    if (focusPeriods[lastIndex].start === now) {
+      return
+    }
+    // close current periods if we miss a blur event
+    const { start } = focusPeriods[lastIndex]
+    focusPeriods[lastIndex] = {
+      start,
+      end: now,
+    }
+  }
+  focusPeriods.push({
+    start: now,
+  })
+}
+
+function closeFocusPeriod() {
+  const lastIndex = focusPeriods.length - 1
+  if (lastIndex >= 0 && focusPeriods[lastIndex].end == null) {
+    const { start } = focusPeriods[lastIndex]
+    focusPeriods[lastIndex] = {
+      start,
+      end: relativeNow(),
+    }
+  }
+}
+
+function trackFocus(onFocusChange: () => void) {
+  return addEventListener(window, DOM_EVENT.FOCUS, () => onFocusChange())
+}
+
+function trackBlur(onBlurChange: () => void) {
+  return addEventListener(window, DOM_EVENT.BLUR, () => onBlurChange())
+}
+
+function getInForeground(startTime: RelativeTime): ForegroundContext {
+  const in_foreground = focusPeriods.some(
+    (focus) => startTime > focus.start && (focus.end == null || startTime < focus.end)
+  )
+  return { view: { in_foreground } }
+}
+
+function getInForegroundPeriods(eventStartTime: RelativeTime, duration: ServerDuration): InForegroundPeriod[] {
+  const eventEndTime = ((eventStartTime as number) + (toDuration(duration) as number)) as RelativeTime
+  return focusPeriods
+    .filter((focus) => {
+      const eventEndsBeforeFocusStart = eventEndTime < focus.start
+      const eventStartsAfterFocusEnds = focus.end == null || eventStartTime > focus.end
+
+      return !eventEndsBeforeFocusStart || !eventStartsAfterFocusEnds
+    })
+    .map((focusPeriod) => {
+      const startTime = eventStartTime > focusPeriod.start ? eventStartTime : focusPeriod.start
+      const startDuration = elapsed(eventStartTime, startTime)
+      const endTime = focusPeriod.end == null || eventEndTime < focusPeriod.end ? eventEndTime : focusPeriod.end
```

Comment:
`focusPeriod.end` cannot be null, could you use `focusePeriod.end === undefined` instead?

Reply:
I like to use `!= null` && `==null` to cover all cases but I do not have a strong opinion, I guess with typescript, the typechecker will yell if null is a possibility and it will be okay.